### ISSUE TITLE
Unable to load PGP, form validation

### ISSUE THREAD

**Issue by dfranzoni** [globaleaks/GlobaLeaks#2218]
Hi,
I cannot load PGP public key in user prefereces.
When I copy and paste or select file with public key the form field are red rounded for validation error and I can't save preferences.

Is it a bug of the new update?

Thank you very much for your help.
**GlobaLeaks version:**
3.0.10

**Browser:**
any

**Comment by evilaliv3** [comment]
Hello @dfranzoni,

the systems accepts public PGP keys in ASCII Armored Format,

i invite you to verify the following:
- you should be uploading a public PGP key
- the format of the key should by ASCII Armored and include the header/trailer.

e.g. and example of valid key is the following: https://www.evilaliv3.org/0x67A0F187.asc

**Comment by dfranzoni** [comment]
Hello Giovanni,
I try your public key but it not work.
I have the same problem.
And my public key is the same.

Sample:

-----BEGIN PGP PUBLIC KEY BLOCK-----
Comment: ID utente: Mario Rossi <mario.rossi@ckube.it>
Comment: Creazione: 14/03/2018 15:18
Comment: Tipo: RSA a 2048 bit (chiave segreta disponibile)
Comment: Uso: Firma, Cifratura, Certificazione degli ID utente
Comment: Impronta digitale: C3A31BC71749EE70CC16771B5BADB0B5B63A684F


mQENBFqpLyYBCADIglNYxrcngkeQjWquCrDsi0yvPxSSs+tnCoS6MTq+40NWGfW7
BAznQnEdZwQafLgRwjPTZh9kES+6eJ6xc7EfT2fSmdM9wLj14XzR6HEOGrYKpI7N
zYzYaM0mXOV/Honr/0NVshGTlDyhLy07UFoHaZBKP6EAHHueWI+UBWv9EHHuvDYV
kio7cwtwO7PRQLFNyCmr5eUuOiAZlgUCSdiq3cl1sI6XV5LoBaVSFOoizwcNXq1Y
b/HZ8JiAvbrSC2lyrYmrLuQ6/b8joco2xAUTYolr4ql7OT5Ve8z00dBqhgDu0ZHE
mrq09q8urWGlALHs3dMcN27neXl94QXGoNbRABEBAAG0Ik1hcmlvIFJvc3NpIDxt
YXJpby5yb3NzaUBja3ViZS5pdD6JAU4EEwEIADgWIQTDoxvHF0nucMwWdxtbrbC1
tjpoTwUCWqkvJgIbAwULCQgHAgYVCgkICwIEFgIDAQIeAQIXgAAKCRBbrbC1tjpo
T8fnB/9WbqwkWePaaUaY4gQMu8Jhs/uuDLMYZbm4JpZtW7aWXh9VOlsLgdL1cRyP
ySgOSzgFj1aRBLQrUu96QDBosOCYUnGy1neBW0ZENgZITTjXoXH8VkKfTYNOxgsm
nkjBnlMhFSGXna+cSH9NWzpy/OIRqMDjXZQz4I/jC4YynsZpl06+ey4ImgHlsbjU
5c7PEmvy4XVj+BMqxK/etKwu1vha74LsPVqwg+F6Y/QhDea387R9oEGU4Y34SY3g
JKWDtkMNRkuzvhuxyEb9BBe0mBpXW7voIJgXggzUujRAvDKfa1wC+WltNCD22SP5
dP1AInP4ybco+rRdeZ+1F3f2+2SUuQENBFqpLyYBCADbnK/c86Rd+RFxuK+VtmG8
PL4vT1Yc9LCw5DbI8bct1EmxmirxiRzD0VeHX7XfypxNEDi2/wHkgUm/DhN/JIew
d+POytjpvflaFuAc/Lx8+5aWpA54VVdEESf7+XNt0Y0BnQfRmm4ou3eruKzcsdWW
FAW+oW4rr5vtyEXOFtTh49NoeDVIfVh5fTQG8JmZsUSRvBPIwUQiEYiGzYMZ93Yw
qJ2xAc9gVnqrsIpfZloXfrEZ/9kRjGZEHiCNSUZd9hsKNFbfqmHQ4PDuDrf113XX
XESVNreU11HCkpGxqLx18mOxejbZmac3e/7NSV51iooUG/I0Hnwu2pXZAokXGyiX
ABEBAAGJATYEGAEIACAWIQTDoxvHF0nucMwWdxtbrbC1tjpoTwUCWqkvJgIbDAAK
CRBbrbC1tjpoT4CNB/4z+qaYuCczyn48KM8VvJPSIE9lVVcq0BfsAIlIZDOnT8l/
eviURDRAAVAJZ/1Td/mzXoS2Hp+GGFwYEWKzwRrYQPqDXKC/N9dXNWL1amkwGEh4
mkE+gbOdp/TgRdbqp51qNDRDQN2WSxAUV04e/AfLIl4qbubKKorSw5zHw3FOhGKw
QoPxN+3z51gR861+76sYCCC4/CQOMmMwfdJdiMLK2H6B56rpWt3nB8/p0Z8MZxqx
xTiKjk05/+r9RPzIb6fMwm+3L4lul6yXmOXaU8jriBKIQsSOR/kptaiO4jQ0GwSL
5atxdVvSO8zfZ4Q6RwoyX6f/45j3ceRnXB2SsilK
=IcJc
-----END PGP PUBLIC KEY BLOCK-----



Davide Franzoni
dfranzoni@coopcsc.it<mailto:dfranzoni@coopcsc.it>
Cell.: 3357987900
Skype: franzdav

[cid:image001.gif@01C73E38.BB0E5D80]
CSC Società Cooperativa Sociale
Via Padre Marcolini,7 - 25040 Ceto (BS)
Tel.: 0364/436406 - Fax: 0364/436407
www.coopcsc.it<http://www.coopcsc.it/> – info@coopcsc.it<mailto:info@coopcsc.it>

Da: Giovanni Pellerano [mailto:notifications@github.com]
Inviato: mercoledì 14 marzo 2018 16:50
A: globaleaks/GlobaLeaks <GlobaLeaks@noreply.github.com>
Cc: Davide Franzoni <dfranzoni@coopcsc.it>; Mention <mention@noreply.github.com>
Oggetto: Re: [globaleaks/GlobaLeaks] Unable to load PGP, form validation (#2218)


Hello @dfranzoni<https://github.com/dfranzoni>,

the systems accepts public PGP keys in ASCII Armored Format,

i invite you to verify the following:

  *   you should be uploading a public PGP key
  *   the format of the key should by ASCII Armored and include the header/trailer.

e.g. and example of valid key is the following: https://www.evilaliv3.org/0x67A0F187.asc

**Comment by evilaliv3** [comment]
Thank your @dfranzoni, i will try to test with your key and get back to you

**Comment by evilaliv3** [comment]
Dear @dfranzoni,

i tried to load your key and i am not able to reproduce your issues.

On which browser are you currently testing?

Here an image during the loading of the key:
![image](https://user-images.githubusercontent.com/217034/37415687-4e1f4396-27ac-11e8-99e5-56ff63f48edc.png)

Here an image of the interface after the key is loaded:
![image](https://user-images.githubusercontent.com/217034/37414843-5e67464c-27aa-11e8-8ff2-c4b5dd5e5d7e.png)

**Comment by dfranzoni** [comment]
Dear Giovanni,
this is my screen:

[cid:image001.png@01D3BBBB.0888D520]

There are some settings that can create this validation problem?

Davide Franzoni
dfranzoni@coopcsc.it<mailto:dfranzoni@coopcsc.it>
Cell.: 3357987900
Skype: franzdav

[cid:image001.gif@01C73E38.BB0E5D80]
CSC Società Cooperativa Sociale
Via Padre Marcolini,7 - 25040 Ceto (BS)
Tel.: 0364/436406 - Fax: 0364/436407
www.coopcsc.it<http://www.coopcsc.it/> – info@coopcsc.it<mailto:info@coopcsc.it>

Da: Giovanni Pellerano [mailto:notifications@github.com]
Inviato: mercoledì 14 marzo 2018 17:24
A: globaleaks/GlobaLeaks <GlobaLeaks@noreply.github.com>
Cc: Davide Franzoni <dfranzoni@coopcsc.it>; Mention <mention@noreply.github.com>
Oggetto: Re: [globaleaks/GlobaLeaks] Unable to load PGP, form validation (#2218)


Dear @dfranzoni<https://github.com/dfranzoni>,

i tried to load your key and i am not able to reproduce your issues.

On which browser are you currently testing?

Here an image during the loading of the key:
[image]<https://user-images.githubusercontent.com/217034/37415687-4e1f4396-27ac-11e8-99e5-56ff63f48edc.png>

Here an image of the interface after the key is loaded:
[image]<https://user-images.githubusercontent.com/217034/37414843-5e67464c-27aa-11e8-8ff2-c4b5dd5e5d7e.png>

**Comment by evilaliv3** [comment]
Dear @dfranzoni  would you please reattach the image via Github?

**Comment by dfranzoni** [comment]
![screen](https://user-images.githubusercontent.com/29452040/37417289-2bd712ec-27b0-11e8-883d-420ce8f0dad7.png)

**Comment by evilaliv3** [comment]
Dear @dfranzoni 

i continue to not identify issues;

On which operating systme / browser / version are you currently testing?

**Comment by NCommander** [comment]
I've managed to reproduce this on a development install of the current GlobaLeaks head.

Browsers:
Firefox 58 / Windows 7
Chrome 64 / Windows 7

I'm trying to run down the root cause.

**Comment by NCommander** [comment]
Using some "known good" test code from the OpenPGP.js bug tracker from here (https://github.com/openpgpjs/openpgpjs/issues/468); attached below. I *think* I figured out what's happening though not why.  At least on the browser versions I'm trying, OpenPGP.JS is falls over trying to walk the entire history of the key causing getStatus and getKeyExpiration to return the incorrection information. Strangely, despite this, encryption operations appear to work fine.

Specifically, both my key (D2247639) and @evilaliv3 key (67A0F187) have points in their history where the key expired, and was later renewed at a point past their expiration date.

In evilaliv3's case, his key expired on 2017-04-07, and was renewed three days later on 2017-04-10 according to the self-sign packets. OpenPGP reports that his key expired on "Fri Apr 07 2017 09:10:55 GMT-0300 (Atlantic Daylight Time)" which is the gap in the self-signed packets.

For my key, the situation is more muddy. I've had a few times where my key has expired and then been renewed after the fact, but none directly corresponding to the reported expiration date of "Wed May 21 2014 18:45:55 GMT-0300 (Atlantic Daylight Time)", I do have a renewal close to that time. 

As part of debugging this, I ran OpenPGP.JS test cases in the browser and got all passes. I'm not confident that my dig into this is correct as right now, GlobaLeaks still rejects a freshly generated key, even though OpenPGP.JS returns status 3 (valid) for it. I've posted my test code so someone can look over what's going on here. I'll look more into it once I've had a solid night's sleep.
[gpg_test_code.zip](https://github.com/globaleaks/GlobaLeaks/files/1814378/gpg_test_code.zip)

**Comment by evilaliv3** [comment]
@sanjanarajan @bartbutler: do you have any hint on this?

the cause may be in a bug affecting just Windows? (i'm not able to reproduce the error on Linux at all)

**Comment by evilaliv3** [comment]
our validator implements the following logic:

```javascript
.factory('glbcKeyLib', ['$q', 'pgp', function($q, pgp) {
    return {
      /**
       * @decription checks to see if passed text is an ascii armored GPG
       * public key. If so, the fnc returns true.
       * @param {String} textInput
       * @return {Bool}
       */
      validPublicKey: function(textInput) {
        if (typeof textInput !== 'string') {
          return false;
        }

        var s = textInput.trim();
        if (s.substr(0, 5) !== '-----') {
          return false;
        }

        // Try to parse the key.
        var result;
        try {
          result = pgp.key.readArmored(s);
        } catch (err) {
          return false;
        }

        // Assert that the parse created no errors.
        if (angular.isDefined(result.err)) {
          return false;
        }

        // Assert that there is only one key in the input.
        if (result.keys.length !== 1) {
          return false;
        }

        var key = result.keys[0];

        // Assert that the key type is not private and the public flag is set.
        if (!key.isPublic() || key.isPrivate()) {
          return false;
        }

        // Verify expiration, revocation, and self sigs.
        if (key.verifyPrimaryKey() !== pgp.enums.keyStatus.valid) {
          return false;
        }

        return true;
      },
    };
}]);
```

**Comment by NCommander** [comment]
I'm now "functional" after a few hours of sleep. Specifically, it's the final test (key.verifyPrimaryKey()) that fails, and looking at the code, it checks the expiration date which is returning non-sense for my key. After futzing with permissions on tor and getting the hidden service to start, I retried on Tor Browser to no avail.

I'll be able to test on a Linux based setup when I get back home tonight, and I'm floating on the IRC channel. I realize I didn't upload my key last night for confirmation that other people can successfully enroll it.

As a confirmation test against "known good", is there a public instance the git HEAD that I can sign into and try uploading a key to? I'm floating on IRC.

[gpg_pubkey.txt](https://github.com/globaleaks/GlobaLeaks/files/1815719/gpg_pubkey.txt)

**Comment by bartbutler** [comment]
I think there was bug in the key expiration time that was just fixed. Can you try the openpgpjs head? Will try to get a release out (3.0.1) today as well.

**Comment by bartbutler** [comment]
I just release 3.0.1. If you are using 3.0.0, please try 3.0.1 and let me know if your issue is not resolved.

**Comment by evilaliv3** [comment]
Thank you @bartbutler.

Was that a bug affecting only windows? because on linux and 3.0.0 i'm not able to reproduce the bug.

Can you point me to the patch to which you are referring?

**Comment by bartbutler** [comment]
It was this one: https://github.com/openpgpjs/openpgpjs/commit/b4055f494ffcd8fe02acdf0447d2772a8f5a84f5

The practical effect of that bug should have just been to extend key expirations further in the future. Not sure if that solves the problem here or not.

**Comment by NCommander** [comment]
I got home from Puerto Rico a few hours ago, and I can confirm 3.0.1 does NOT fix the issue for me.

I also setup a clean GlobaLeaks development instance on my desktop computer (which runs Linux) and still remain unable to import keys on Chrome/Firefox for Linux. I do question if this MIGHT be an issue that both of my dev instances are running on Gentoo vs. Ubuntu but given all the validation code is client-side, I can't see how that can affect it. As I've got a more capable machine, I'm working through my existing testing code in an effort to localize the specific issue here.

**Comment by NCommander** [comment]
I decided to futz with this a bit more before sleeping off the end of my trip.

@bartbutler 

Stepping through openpgpjs, I think the problem is getPrimaryUser; it's grabbing the wrong key. I haven't figured out a good way to attach a debugger to the OpenPGP source so I'm mostly examining objects with Chrome (any tips here would be welcome; I've only managed to use the debugger with the post-processed babel source which ... less than ideal).

Take a look at the object here which lists the key.user.selfSignatures. The most recent signature is in slot 1 which extends the key validity although there are multiple signatures before and after it.

![image](https://user-images.githubusercontent.com/181056/37505174-c415b27a-28b9-11e8-8cc8-ae8b2ffeb80a.png)

getPrimaryUser isn't grabbing the correct key.

![image](https://user-images.githubusercontent.com/181056/37505754-c553e6cc-28bc-11e8-966f-cc895d4379df.png)

Which causes the expired check in verifyPrimaryKey to go bang. I'm not sure why I'm encountering this issue and @evilaliv3 isn't but I started implemented a basic test case with this public key running in node, and I get 1 (expired) here which is consistent with what I'm getting on Chrome/Firefox.

**Comment by evilaliv3** [comment]
Thank you @NCommander 

I just retried on a clean install of Widows with Edge and i'm not able to reproduce the bug with all the keys that have been pasted here.

Maybe thanks to your feedback @bartbutler is now able to look at it in detail.

**Comment by NCommander** [comment]
Testing with 3.0.2, all the test keys I tried now return the correct expiration time with my test code \o/

However, I still can't import the key into GlobaLeaks. as this bit still evaluates to false.
```
        if (key.verifyPrimaryKey() !== pgp.enums.keyStatus.valid) {
          return false;
        }
````

verifyPrimaryKey is an async function. How does Angular resolve the promise here? In my JavaScript test code, I had to do the following to test the same functionality.

```        
key.verifyPrimaryKey().then(keyStatus => {
            if ( keyStatus !== openpgp.enums.keyStatus.valid) {
                console.log("Key Invalid");
            }
        })
```

As a further test, I rewrote the function in main.js in my test code to this:
```
        // Verify expiration, revocation, and self sigs.
//        if (key.verifyPrimaryKey() !== pgp.enums.keyStatus.valid) {
//          return false;
//        }

        key.verifyPrimaryKey().then(keyStatus => {
            console.log("Evaluating key status");
            if (keyStatus !== pgp.enums.keyStatus.valid) {
                console.log("keystatus invalid");
                return false;
            }
        });
```

Which prints "Evaluatingkey status" twice in the console.log and succeeds. Given this has worked up to this point, I can't help but feel like my fix is bogus, but I don't know enough about Angular and how promises are resolved in this case to determine what's going on.

**Comment by sanjanarajan** [comment]
Your fix evaluates the promise correctly, but using ES7 syntax you can simplify the if statement to
` if (await key.verifyPrimaryKey() !== pgp.enums.keyStatus.valid)`

**Comment by NCommander** [comment]
I'll get a PR up in a bit with this, and the requirements.txt/package.json changes for review.

**Comment by sanjanarajan** [comment]
I'm not sure how you are currently generating this list of user IDs, but if you want to know if a given user is valid, you can call the User.prototype.verify function on that user. I tried it on your key and correctly got a result of 2 (enums.keyStatus.revoked) for the michael.casadevall@canonical.com user.

**Comment by NCommander** [comment]
I pushed a PR that appears to properly fix this, although there are still lingering issues with how the keys display and such.

The example code above doesn't properly handle async/await and was erroneously returning true when it should have returned false due to the .then branch not being immediately and falling through. The PR I opened changes the validator to be async and operates as a Promise which correctly allows the OpenPGP.js functions to work. There are still open issues as this doesn't fix the displayed key info once the key is enrolled due to the same async function/non-async loader.

I've got essentially zip experience with Angular, so this patch was written with a lot of Googling and referencing Angular documentation. I did test this with my personal key, and an old expired one, and I got the expected behavior for both. Part of me does wonder if it would be better to show why validation failed and the Save button is grayed out.
